src/components/tl/mlx5/tl_mlx5_dm.c

```diff
@@ -0,0 +1,179 @@
+            min_chunks_to_alloc = 1;
+            if (!max_chunks_to_alloc) {
+                tl_error(lib,
+                         "requested buffer size (=%ld) is too large, "
+                         "should be set to be strictly less than %ld. "
+                         "max allocation size is %ld",
+                         buf_size, attr.max_dm_size / 2, attr.max_dm_size);
+                return UCC_ERR_NO_RESOURCE;
+            }
+        } else {
+            max_chunks_to_alloc = min_chunks_to_alloc = *buf_num_p;
+        }
+        if (attr.max_dm_size < buf_size * min_chunks_to_alloc) {
+            tl_error(lib,
+                     "cannot allocated %i buffer(s) of size %ld, "
+                     "max allocation size is %ld",
+                     min_chunks_to_alloc, buf_size, attr.max_dm_size);
+            return UCC_ERR_NO_MEMORY;
+        }
+        for (i = max_chunks_to_alloc; i >= min_chunks_to_alloc; i--) {
+            dm_attr.length = i * buf_size;
+            errno          = 0;
+            dm_ptr         = ibv_alloc_dm(ib_ctx, &dm_attr);
+            if (dm_ptr) {
+                break;
+            }
+        }
+        if (!dm_ptr) {
+            tl_error(lib,
+                     "dev mem allocation failed, requested %ld, attr.max %zd, "
+                     "errno %d",
+                     dm_attr.length, attr.max_dm_size, errno);
+            return errno == ENOMEM || errno == ENOSPC ? UCC_ERR_NO_MEMORY
+                                                      : UCC_ERR_NO_MESSAGE;
+        }
+        dm_mr = ibv_reg_dm_mr(pd, dm_ptr, 0, dm_attr.length,
+                              IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE |
+                                  IBV_ACCESS_ZERO_BASED);
+        if (!dm_mr) {
+            tl_error(lib, "failed to reg memic");
+            ibv_free_dm(dm_ptr);
+            return UCC_ERR_NO_MESSAGE;
+        }
+        *buf_num_p = i;
+    }
+    *ptr = dm_ptr;
+    *mr  = dm_mr;
+
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_mlx5_dm_init(ucc_tl_mlx5_team_t *team)
+{
+    ucc_tl_mlx5_context_t *   ctx = UCC_TL_MLX5_TEAM_CTX(team);
+    ucc_tl_mlx5_lib_config_t *cfg = &UCC_TL_MLX5_TEAM_LIB(team)->cfg;
+    ucc_status_t              status;
+
+    status = ucc_tl_mlx5_dm_alloc_reg(
+        ctx->shared_ctx, ctx->shared_pd, cfg->dm_host, cfg->dm_buf_size,
+        &cfg->dm_buf_num, &team->dm_ptr, &team->dm_mr, UCC_TL_TEAM_LIB(team));
+    if (status != UCC_OK) {
+        tl_error(UCC_TL_TEAM_LIB(team),
+                 "failed to alloc and register device memory");
+        return status;
+    }
+    team->oob_req = NULL;
+
+    // TODO: fix case dm_host=true
+    status = ucc_mpool_init(&team->dm_pool, 0, sizeof(ucc_tl_mlx5_dm_chunk_t),
+                            0, UCC_CACHE_LINE_SIZE, cfg->dm_buf_num,
+                            cfg->dm_buf_num, &ucc_tl_mlx5_dm_ops,
+                            UCC_THREAD_MULTIPLE, "mlx5 dm pool");
```

Comment:
pls use actualy thread mode instead of UCC_THREAD_MULTIPLE

Reply:
can you confirm it is ok ?